waspc/packages/wasp-config/__tests__/mapUserSpecToAppSpecDecls.unit.test.ts

```diff
@@ -0,0 +1,1103 @@
+      }
+    }
+    const entityRefParser = makeRefParser('Entity', entities)
+    const routeRefParser = makeRefParser('Route', routes)
+
+    const result = mapApp(
+      userSpec.app.config,
+      entityRefParser,
+      routeRefParser,
+      userSpec.auth,
+      userSpec.server,
+      userSpec.client,
+      userSpec.db,
+      userSpec.emailSender,
+      userSpec.websocket
+    )
+
+    expect(result).toStrictEqual({
+      wasp: {
+        version: userSpec.app.config.wasp.version,
+      },
+      title: userSpec.app.config.title,
+      head: userSpec.app.config.head,
+      auth:
+        userSpec.auth &&
+        mapAuth(userSpec.auth, entityRefParser, routeRefParser),
+      server: userSpec.server && mapServer(userSpec.server),
+      client: userSpec.client && mapClient(userSpec.client),
+      db: userSpec.db && mapDb(userSpec.db),
+      emailSender: userSpec.emailSender && mapEmailSender(userSpec.emailSender),
+      webSocket: userSpec.websocket && mapWebSocket(userSpec.websocket),
+    } satisfies AppSpec.App)
+  }
+})
+
+describe('mapAuth', () => {
+  test('should map minimal config correctly', () => {
+    testMapAuth(Fixtures.getAuth('minimal'))
+  })
+
+  test('should map full config correctly', () => {
+    testMapAuth(Fixtures.getAuth('full'))
+  })
+
+  test('should throw if userEntity is not provided to entity parser', () => {
+    testMapAuth(Fixtures.getAuth('minimal'), {
+      overrideEntities: [],
+      shouldError: true,
+    })
+  })
+
+  test('should throw if externalAuthEntity ref is not provided when defined', () => {
+    const auth = Fixtures.getAuth('full')
+    testMapAuth(auth, {
+      overrideEntities: [auth.userEntity],
+      shouldError: true,
+    })
+  })
+
+  test('should throw if emailVerification clientRoute ref is not provided when defined', () => {
+    const auth = Fixtures.getAuth('full')
+    testMapAuth(auth, {
+      overrideRoutes: [Fixtures.getRoute('password-reset').name],
+      shouldError: true,
+    })
+  })
+
+  test('should throw if passwordReset clientRoute ref is not provided when defined', () => {
+    const auth = Fixtures.getAuth('full')
+    testMapAuth(auth, {
+      overrideRoutes: [Fixtures.getRoute('email-verification').name],
+      shouldError: true,
+    })
+  })
+
+  function testMapAuth(
+    auth: UserApi.AuthConfig,
+    options:
+      | {
+          overrideEntities?: string[]
+          overrideRoutes?: string[]
+          shouldError: boolean | undefined
+        }
+      | undefined = {
+      overrideEntities: undefined,
+      overrideRoutes: undefined,
+      shouldError: false,
+    }
+  ): void {
+    const { overrideEntities, overrideRoutes, shouldError } = options
+    const entities = overrideEntities ?? []
+    if (!overrideEntities) {
+      if (auth.userEntity) {
+        entities.push(auth.userEntity)
+      }
+      if (auth.externalAuthEntity) {
+        entities.push(auth.externalAuthEntity)
+      }
+    }
```

Comment:
This threw me off quite a bit due to inverse logic of override entities instead being overriden by default entities. What about this:

```suggestion
    const entities = overrideEntities ?? _.compact([auth.userEntity, auth.externalAuthEntity]);
```

Same for routes below.

Reply:
We don't have `lodash` in this proejct 😄

Reply:
Yeah or you can even just do `.filter(Boolean)`, I think we want only truthy values here anyway.

Also feel free to import lodash, it is useful, and we use it in other projects already.

Reply:
I can't since `Boolean` doesn't type narrow correctly. Resulting array us `(string | undefined)[]`. Did that initially though. I thought typescript 5.5 fixed that, but it was something else with `filter()`.